mail_server_relay_disallowed/models/ir_mail_server.py

```diff
@@ -0,0 +1,109 @@
+                          "any outbound email"
+
+        # The email's "Envelope From" (Return-Path), and all recipient
+        # addresses must only contain ASCII characters.
+        from_rfc2822 = extract_rfc2822_addresses(smtp_from)
+        assert from_rfc2822, ("Malformed 'Return-Path' or 'From' address: "
+                              "%r - "
+                              "It should contain one valid plain ASCII "
+                              "email") % smtp_from
+        # use last extracted email, to support rarities like 'Support@MyComp
+        # <support@mycompany.com>'
+        smtp_from = from_rfc2822[-1]
+        email_to = message['To']
+        email_cc = message['Cc']
+        email_bcc = message['Bcc']
+        del message['Bcc']
+
+        smtp_to_list = [
+            address
+            for base in [email_to, email_cc, email_bcc]
+            for address in extract_rfc2822_addresses(base)
+            if address
+        ]
+        assert smtp_to_list, self.NO_VALID_RECIPIENT
+
+        x_forge_to = message['X-Forge-To']
+        if x_forge_to:
+            # `To:` header forged, e.g. for posting on mail.channels,
+            # to avoid confusion
+            del message['X-Forge-To']
+            del message['To']  # avoid multiple To: headers!
+            message['To'] = x_forge_to
+
+        # Do not actually send emails in testing mode!
+        if getattr(threading.currentThread(), 'testing',
+                   False) or self.env.registry.in_test_mode():
+            _test_logger.info("skip sending email in test mode")
+            return message['Message-Id']
+
+        try:
+            message_id = message['Message-Id']
+
+            # START OF CODE ADDED
+            smtp = self.connect(
+                smtp_server, smtp_port, smtp_user, smtp_password,
+                smtp_encryption or False, smtp_debug
+            )
+
+            from email.utils import parseaddr, formataddr
+            # exact name and address
+            (oldname, oldemail) = parseaddr(message['From'])
+            # use original name with new address
+            newfrom = formataddr((oldname, smtp.user))
+            # need to use replace_header instead '=' to prevent
+            # double field
+            message.replace_header('From', newfrom)
+            smtp.sendmail(smtp.user, smtp_to_list, message.as_string())
+            # END OF CODE ADDED
+
+            # do not quit() a pre-established smtp_session
+            if not smtp_session:
+                smtp.quit()
+        except Exception as e:
+            msg = _(
+                "Mail delivery failed via SMTP server '%s'.\n%s: %s"
+            ) % (
+                tools.ustr(smtp_server),
+                e.__class__.__name__,
+                tools.ustr(e)
+            )
+            _logger.error(msg)
+            raise MailDeliveryException(_("Mail Delivery Failed"), msg)
+        return message_id
```

Comment:
I think here it should adapt changes done in original v13 code.

Reply:
done